### DISCUSSION TITLE
Is it possible to make greptimedb upon databend？

### DISCUSSION THREAD

**Discussion by sundy-li** [GreptimeTeam/greptimedb#1150] [Ideas]
Hello, greptimedb members.

I noticed that greptimedb's query engine is powered by datafusion, which it's similar way like `influxdb-iox`.

Databend has a high-performance computing engine based on the Arrow memory model.  I saw that greptimdb is depended on opendal and its type system and expression referred to the old version of databend's. 

So I am wondering if we can build the computation layer of greptimedb on top of it.

Advantages:

1. We have a high-performance computing engine with `expression`, `function`, and `aggregate function` support, reference to [clickbench](https://benchmark.clickhouse.com/#eyJzeXN0ZW0iOnsiQXRoZW5hIChwYXJ0aXRpb25lZCkiOnRydWUsIkF0aGVuYSAoc2luZ2xlKSI6dHJ1ZSwiQXVyb3JhIGZvciBNeVNRTCI6dHJ1ZSwiQXVyb3JhIGZvciBQb3N0Z3JlU1FMIjp0cnVlLCJCeXRlSG91c2UiOnRydWUsIkNpdHVzIjp0cnVlLCJjbGlja2hvdXNlLWxvY2FsIChwYXJ0aXRpb25lZCkiOnRydWUsImNsaWNraG91c2UtbG9jYWwgKHNpbmdsZSkiOnRydWUsIkNsaWNrSG91c2UiOnRydWUsIkNsaWNrSG91c2UgKHR1bmVkKSI6dHJ1ZSwiQ2xpY2tIb3VzZSAoenN0ZCkiOnRydWUsIkNsaWNrSG91c2UgQ2xvdWQiOnRydWUsIkNyYXRlREIiOnRydWUsIkRhdGFiZW5kIjp0cnVlLCJEYXRhRnVzaW9uIChzaW5nbGUpIjp0cnVlLCJBcGFjaGUgRG9yaXMiOnRydWUsIkRydWlkIjp0cnVlLCJEdWNrREIgKFBhcnF1ZXQpIjp0cnVlLCJEdWNrREIiOnRydWUsIkVsYXN0aWNzZWFyY2giOnRydWUsIkVsYXN0aWNzZWFyY2ggKHR1bmVkKSI6dHJ1ZSwiR3JlZW5wbHVtIjp0cnVlLCJIZWF2eUFJIjp0cnVlLCJIeWRyYSI6dHJ1ZSwiSW5mb2JyaWdodCI6dHJ1ZSwiS2luZXRpY2EiOnRydWUsIk1hcmlhREIgQ29sdW1uU3RvcmUiOnRydWUsIk1hcmlhREIiOnRydWUsIk1vbmV0REIiOnRydWUsIk1vbmdvREIiOnRydWUsIk15U1FMIChNeUlTQU0pIjp0cnVlLCJNeVNRTCI6dHJ1ZSwiUGlub3QiOnRydWUsIlBvc3RncmVTUUwiOnRydWUsIlBvc3RncmVTUUwgKHR1bmVkKSI6dHJ1ZSwiUXVlc3REQiAocGFydGl0aW9uZWQpIjp0cnVlLCJRdWVzdERCIjp0cnVlLCJSZWRzaGlmdCI6dHJ1ZSwiU2VsZWN0REIiOnRydWUsIlNpbmdsZVN0b3JlIjp0cnVlLCJTbm93Zmxha2UiOnRydWUsIlNRTGl0ZSI6dHJ1ZSwiU3RhclJvY2tzIjp0cnVlLCJUaW1lc2NhbGVEQiAoY29tcHJlc3Npb24pIjp0cnVlLCJUaW1lc2NhbGVEQiI6dHJ1ZX0sInR5cGUiOnsic3RhdGVsZXNzIjp0cnVlLCJtYW5hZ2VkIjp0cnVlLCJKYXZhIjp0cnVlLCJjb2x1bW4tb3JpZW50ZWQiOnRydWUsIkMrKyI6dHJ1ZSwiTXlTUUwgY29tcGF0aWJsZSI6dHJ1ZSwicm93LW9yaWVudGVkIjp0cnVlLCJDIjp0cnVlLCJQb3N0Z3JlU1FMIGNvbXBhdGlibGUiOnRydWUsIkNsaWNrSG91c2UgZGVyaXZhdGl2ZSI6dHJ1ZSwiZW1iZWRkZWQiOnRydWUsIlJ1c3QiOnRydWUsInNlYXJjaCI6dHJ1ZSwiZG9jdW1lbnQiOnRydWUsInRpbWUtc2VyaWVzIjp0cnVlfSwibWFjaGluZSI6eyJzZXJ2ZXJsZXNzIjpmYWxzZSwiMTZhY3UiOmZhbHNlLCJMIjpmYWxzZSwiTSI6ZmFsc2UsIlMiOmZhbHNlLCJYUyI6ZmFsc2UsImM2YS40eGxhcmdlLCA1MDBnYiBncDIiOnRydWUsImM1LjR4bGFyZ2UsIDUwMGdiIGdwMiI6ZmFsc2UsImM2YS5tZXRhbCwgNTAwZ2IgZ3AyIjpmYWxzZSwiMTYgdGhyZWFkcyI6ZmFsc2UsIjIwIHRocmVhZHMiOmZhbHNlLCIyNCB0aHJlYWRzIjpmYWxzZSwiMjggdGhyZWFkcyI6ZmFsc2UsIjMwIHRocmVhZHMiOmZhbHNlLCI0OCB0aHJlYWRzIjpmYWxzZSwiNjAgdGhyZWFkcyI6ZmFsc2UsIm01ZC4yNHhsYXJnZSI6ZmFsc2UsImM2YS40eGxhcmdlLCAxNTAwZ2IgZ3AyIjpmYWxzZSwicmEzLjE2eGxhcmdlIjpmYWxzZSwicmEzLjR4bGFyZ2UiOmZhbHNlLCJyYTMueGxwbHVzIjpmYWxzZSwiUzI0IjpmYWxzZSwiUzIiOmZhbHNlLCIyWEwiOmZhbHNlLCIzWEwiOmZhbHNlLCI0WEwiOmZhbHNlLCJYTCI6ZmFsc2V9LCJjbHVzdGVyX3NpemUiOnsiMSI6dHJ1ZSwiMiI6dHJ1ZSwiNCI6dHJ1ZSwiOCI6dHJ1ZSwiMTYiOnRydWUsIjMyIjp0cnVlLCI2NCI6dHJ1ZSwiMTI4Ijp0cnVlLCJzZXJ2ZXJsZXNzIjp0cnVlLCJ1bmRlZmluZWQiOnRydWV9LCJtZXRyaWMiOiJob3QiLCJxdWVyaWVzIjpbdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZSx0cnVlLHRydWUsdHJ1ZV19). greptimedb can focus on optimizing for time-series scenarios.

2. `Greptimedb` can reuse the `pipeline` framework of databend.

Disadvantages:

1. Databend may need to spilt the computing engine into a library for greptimedb to call.
2. Some code migration and refactoring work may be required in greptimedb

**Comment by waynexia** [comment]
Hi!

This is a very interesting idea and the performance of databend is very impressive according to the report 🚀 .

As you said, both databend and datafusion have memory models based on Arrow, so there is not much difference between us. But I'm not familiar with databend's execution engine, so I don't know how much it differs from datafusion in terms of functionality and data type support.

And, although we have abstracted the execution engine, there are still many places where we are tied to DataFusion, so switching (or adding support for the second engine) would require a lot of work for both parties. But I'm very optimistic about this proposal. Let's keep communicating on the details ❤️ !

**Comment by MichaelScofield** [comment]
The query engine and pipeline framework in Databend look promising. If we were to adopt them, where to start first? @sundy-li

**Reply by sundy-li** [reply]
We may need to separate the pipeline framework to be a standalone crate.

**Reply by MichaelScofield** [reply]
Looking forward to it!

**Comment by evenyag** [comment]
Cool! I'm very excited about this. @sundy-li 

## GreptimeDB
There are some blockers that greptimedb needs to resolve or workaround:
1. Its type system and some crates are built upon the official arrow instead of arrow2. But I'm happy to see that the arrow community is going to unify these two crates jorgecarleitao/arrow2#1429
2. Its promql extensions, and script engine are highly coupled with datafusion
3. We must both replace the type system and the query engine due to the first point as the underlying arrow crates are incompatible

## Databend
There are also some points that databend might need to consider:
1. Is it possible to provide a way to allow users to define their own logical/physical plans (or even data types)?
2. greptimedb and databend have different catalog systems
4. Is it possible to share the same distributed query framework? DataFusion doesn't provide distributed planning and execution support.
5. It'd be useful to allow users to attach information to the query context (during planning/executing) and access it via downcasting (ergonomic than `HashMap<String, String>`)

## Share People's Efforts
Recently, I'm going to re-design and refactor our type system, including:
- decouple logical data types and physical vector types, so we are able to use different vector implementations to represent the same logical type (e.g. Use DictionaryArray/StringArray for String type)
- implement more data types

I just realize that there are might be some other projects repeating the same work:
- some might need a simple `Value`/`DataValue` wrapper
- some might need a `Vector`/`Column`
- some might also need an expression evaluation framework

Could we implement something like query engine building blocks, such as the type system, expression framework, and execution framework (like [velox](https://github.com/facebookincubator/velox))? So we could combine people's efforts and eliminate some repeating work? DataFusion tries to solve similar things, but it's highly coupled with the official arrow's API.

## Conclusion
It'd be great if users can have more choices. Maybe extracting databend's type system into an independent library would be a good starting point. I can also help bridge arrow and arrow2.

**Reply by sundy-li** [reply]
> I can also help bridge arrow and arrow2.

We are going to move back into arrow-rs. Let's stay tuned.

>  I'm going to re-design and refactor our type system

A good way to go, we spent nearly 6 months to refactor the type system into a new one.
